In [4]:
import pandas as pd

## Q1. Pandas version

In [5]:
pd.__version__

'2.3.1'

Getting the data

In [7]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv

--2025-10-18 09:02:10--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 874188 (854K) [text/plain]
Saving to: ‘car_fuel_efficiency.csv’

car_fuel_efficiency 100%[===================>] 853.70K  --.-KB/s    in 0.01s   

2025-10-18 09:02:10 (76.6 MB/s) - ‘car_fuel_efficiency.csv’ saved [874188/874188]



In [10]:
df = pd.read_csv('car_fuel_efficiency.csv')

In [11]:
df

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                     170            3.0       159.0     3413.433759   
1                     130            5.0        97.0     3149.664934   
2                     170            NaN        78.0     3079.038997   
3                     220            4.0         NaN     2542.392402   
4                     210            1.0       140.0     3460.870990   
...                   ...            ...         ...             ...   
9699                  140            5.0       164.0     2981.107371   
9700                  180            NaN       154.0     2439.525729   
9701                  220            2.0       138.0     2583.471318   
9702                  230            4.0       177.0     2905.527390   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year  origin fuel_type         drivetrain  \
0             17.7        2003  Europe  Gasoline    All-wheel drive   
1             17.8        2007     USA  Gasoline  Front-wheel drive   
2             15.1        2018  Europe  Gasoline  Front-wheel drive   
3             20.2        2009     USA    Diesel    All-wheel drive   
4             14.4        2009  Europe  Gasoline    All-wheel drive   
...            ...         ...     ...       ...                ...   
9699          17.3        2013  Europe    Diesel  Front-wheel drive   
9700          15.0        2004     USA  Gasoline    All-wheel drive   
9701          15.1        2008     USA    Diesel    All-wheel drive   
9702          19.4        2011     USA    Diesel  Front-wheel drive   
9703          14.7        2005    Asia    Diesel    All-wheel drive   

      num_doors  fuel_efficiency_mpg  
0           0.0            13.231729  
1           0.0            13.688217  
2           0.0            14.246341  
3           2.0            16.912736  
4           2.0            12.488369  
...         ...                  ...  
9699        NaN            15.101802  
9700        0.0            17.962326  
9701       -1.0            17.186587  
9702        1.0            15.331551  
9703       -1.0            14.884467  

[9704 rows x 11 columns]

## Q2. Records count

How many records are in the dataset?

- 4704
- 8704
- 9704
- 17704

Answer: 9704

## Q3. Fuel types

How many fuel types are presented in the dataset?

- 1
- 2
- 3
- 4

Anwswer: 2

In [13]:
df.describe(include='all')

engine_displacement  num_cylinders   horsepower  vehicle_weight  \
count           9704.000000    9222.000000  8996.000000     9704.000000   
unique                  NaN            NaN          NaN             NaN   
top                     NaN            NaN          NaN             NaN   
freq                    NaN            NaN          NaN             NaN   
mean             199.708368       3.962481   149.657292     3001.280993   
std               49.455319       1.999323    29.879555      497.894860   
min               10.000000       0.000000    37.000000      952.681761   
25%              170.000000       3.000000   130.000000     2666.248985   
50%              200.000000       4.000000   149.000000     2993.226296   
75%              230.000000       5.000000   170.000000     3334.957039   
max              380.000000      13.000000   271.000000     4739.077089   

        acceleration   model_year  origin fuel_type       drivetrain  \
count    8774.000000  9704.000000    9704      9704             9704   
unique           NaN          NaN       3         2                2   
top              NaN          NaN  Europe  Gasoline  All-wheel drive   
freq             NaN          NaN    3254      4898             4876   
mean       15.021928  2011.484027     NaN       NaN              NaN   
std         2.510339     6.659808     NaN       NaN              NaN   
min         6.000000  2000.000000     NaN       NaN              NaN   
25%        13.300000  2006.000000     NaN       NaN              NaN   
50%        15.000000  2012.000000     NaN       NaN              NaN   
75%        16.700000  2017.000000     NaN       NaN              NaN   
max        24.300000  2023.000000     NaN       NaN              NaN   

          num_doors  fuel_efficiency_mpg  
count   9202.000000          9704.000000  
unique          NaN                  NaN  
top             NaN                  NaN  
freq            NaN                  NaN  
mean      -0.006412            14.985243  
std        1.048162             2.556468  
min       -4.000000             6.200971  
25%       -1.000000            13.267459  
50%        0.000000            15.006037  
75%        1.000000            16.707965  
max        4.000000            25.967222

In [14]:
df.dtypes.value_counts()

float64    6
object     3
int64      2
Name: count, dtype: int64

In [16]:
df['fuel_type'].nunique()

2

## Q4. Missing values

How many columns in the dataset have missing values?

- 0
- 1
- 2
- 3
- 4

Answer: 4

In [22]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

## Q5. Max fuel efficiency

What's the maximum fuel efficiency of cars from Asia?

- 13.75
- 23.75
- 33.75
- 43.75

Answer: 23.75

In [25]:
df[df['origin'] == 'Asia']['fuel_efficiency_mpg'].max()

23.759122836520497

## Q6. Median value of horsepower

- Yes, it increased
- Yes, it decreased
- No

Answer: Yes, it increased

In [29]:
df.groupby('model_year')['horsepower'].median()


model_year
2000    149.0
2001    149.0
2002    145.0
2003    147.0
2004    147.0
2005    147.5
2006    151.0
2007    151.0
2008    154.0
2009    150.0
2010    148.5
2011    152.0
2012    151.0
2013    148.0
2014    151.0
2015    151.0
2016    150.0
2017    149.0
2018    150.0
2019    148.0
2020    150.5
2021    149.0
2022    149.0
2023    150.5
Name: horsepower, dtype: float64

In [30]:
df['horsepower'].median()

149.0

## Q7. Sum of weights

- 0.051
- 0.51
- 5.1
- 51

Answer: 0.51

In [33]:
df["weight_norm"] = df["vehicle_weight"] / df["vehicle_weight"].sum()
df["weight_norm"].sum()

np.float64(1.0)